### PR TITLE
MM-17087 - Disable plugin on removal

### PR DESCRIPTION
**Pull request by ali-farooq0**
#### Summary
- Disable plugin on removal
- Updated cluster peers synchronization logic to notify clients ONLY after all cluster peers have been fully updated, as opposed to notifying them when one of them was updated
- Re-activates plugin on upgrade via `PluginEnvironment` as opposed to `EnablePlugin` api which caused a config change + notify cluster peer of a config change and wreaked havoc -- true story.
- Added a unit test to confirm `disabled` state on plugin re-install
- Not able to write a unit test to test cluster functionality, but hopefully in the future we can add isolated env to test these cases and lock down behaviour

#### Dev testing
- Wrote a mini stress script that does the following. See attached script to the jira.
 -- Removed plugin 1.0
 -- Installs plugin 1.0
 -- Enables plugin
 -- Installs plugin 2.0
 -- Installs plugin 1.0
 -- repeat
- With the stress test, I was still able to reproduce a 404 bundle not found (1 in ~100 runs), however that was probably caused by my script that had already upgraded to the newer plugin version (as we remove bundle on deactivate/upgrade) while the clients were going and fetching the bundle, not a 100% certain though. That is my theory as i couldn't find any error logs for `missing_plugin`, which is caused when the plugin is missing from an instance.

#### Ticket Link
Fixes https://mattermost.atlassian.net/browse/MM-17087

#### Related Pull Requests
- Has webapp changes (https://github.com/mattermost/mattermost-webapp/pull/3425)

### CODE REVIEW

**Comment by ali-farooq0** [comment]
Tips and tricks to connect to individual node on HA


- Running `nslookup mattermost-server-pr-11779.test.mattermost.cloud` will give you the addresses of each node
- Could also just run 
```
> nslookup mattermost-server-pr-11779.test.mattermost.cloud | grep -e "Address:" |  awk -F'[:]' '{print $2}' | awk '{ sub(/^[ \t]+/, ""); print }' | awk 'NR>1'
34.233.45.132
52.6.190.4
```
- Could also look up the address in devtools
image

File: app/plugin.go [hunk 1]
```diff
@@ -99,10 +99,11 @@ func (a *App) SyncPluginsActiveState() {
...
 				}
 
-				if activated && updatedManifest.HasClient() {
-					message := model.NewWebSocketEvent(model.WEBSOCKET_EVENT_PLUGIN_ENABLED, "", "", "", nil)
-					message.Add("manifest", updatedManifest.ClientManifest())
-					a.Publish(message)
+				if activated {
+					// Notify all cluster clients if ready
+					if err := a.notifyPluginEnabled(updatedManifest); err != nil {
+						a.Log.Error("Unable to notify plugin enabled", mlog.Err(err))
```
Comment (lieut-data) [inline app/plugin.go]: nit:
```suggestion
						a.Log.Error("Failed to notify cluster on plugin enable", mlog.Err(err))

File: app/plugin.go [hunk 2]
```diff
@@ -394,3 +398,48 @@ func (a *App) GetPlugins() (*model.PluginsResponse, *model.AppError) {
 
 	return resp, nil
 }
+
+// notifyPluginEnabled notifies web clients for all cluster peers if
+// manifest version is same across all peers, including self.
+func (a *App) notifyPluginEnabled(manifest *model.Manifest) *model.AppError {
+	if manifest.HasClient() {
```
Comment (lieut-data) [inline app/plugin.go]: nit: let's invert this and early return, to make the happy path less indented

File: app/plugin.go [hunk 3]
```diff
@@ -394,3 +398,48 @@ func (a *App) GetPlugins() (*model.PluginsResponse, *model.AppError) {
...
+		var statuses model.PluginStatuses
+		expectedPeers := 1
```
Comment (lieut-data) [inline app/plugin.go]: nit: I'd expect `expectedPeers := 0`, since intuitively I am not my own `peer`.

File: app/plugin.go [hunk 4]
```diff
@@ -394,3 +398,48 @@ func (a *App) GetPlugins() (*model.PluginsResponse, *model.AppError) {
...
+			// Count consistent peers
+			isPeerConsistent := status.Version == manifest.Version &&
+				localStatus.State == model.PluginStateRunning && localStatus.State == status.State
+			if status.PluginId == manifest.Id && isPeerConsistent {
+				actualPeers++
+			}
+		}
+
+		// Notify all cluster peer clients.
+		if actualPeers == expectedPeers {
```
Comment (lieut-data) [inline app/plugin.go]: Instead of counting, could we flip things around to instead just return if we find a peer whose manifest doesn't match my own? This naturally collapses when there are no peers, and should simplify things tremendously.

File: app/plugin.go [hunk 5]
```diff
@@ -394,3 +398,48 @@ func (a *App) GetPlugins() (*model.PluginsResponse, *model.AppError) {
 
 	return resp, nil
 }
+
+// notifyPluginEnabled notifies web clients for all cluster peers if
```
Comment (lieut-data) [inline app/plugin.go]: Can you go into some extra detail on the thinking behind this method, especially our discussion yesterday on why we can't just notify our own clients (e.g. because they could connect to another host that doesn't have the bundle)? 

Also, could you describe why this is guaranteed to converge, e.g. why two clients won't end up checking each other and skipping the update, leaving no one left to do so?

File: app/plugin_install.go [hunk 6]
```diff
@@ -45,6 +51,10 @@ func (a *App) RemovePluginFromData(data model.PluginEventData) {
 	if err := a.removePluginLocally(data.Id); err != nil {
 		mlog.Error("Failed to remove plugin locally", mlog.Err(err), mlog.String("id", data.Id))
 	}
+
+	if err := a.notifyPluginStatusesChanged(); err != nil {
```
Comment (lieut-data) [inline app/plugin_install.go]: Just looking at the diff at this point in the review, and it's entirely not clear why is a `notifyPluginEvents` and a `notifyPluginStatusesChanged` -- feels like we could converge on fewer methods?

File: app/plugin_install.go [hunk 7]
```diff
@@ -61,40 +71,45 @@ func (a *App) installPlugin(pluginFile io.ReadSeeker, replace bool) (*model.Mani
 	// Store bundle in the file store to allow access from other servers.
 	pluginFile.Seek(0, 0)
 
-	if _, err := a.WriteFile(pluginFile, a.getBundleStorePath(manifest.Id)); err != nil {
-		return nil, model.NewAppError("uploadPlugin", "app.plugin.store_bundle.app_error", nil, err.Error(), http.StatusInternalServerError)
+	if _, appErr := a.WriteFile(pluginFile, a.getBundleStorePath(manifest.Id)); appErr != nil {
+		return nil, model.NewAppError("uploadPlugin", "app.plugin.store_bundle.app_error", nil, appErr.Error(), http.StatusInternalServerError)
 	}
 
+	// Notify cluster peers async.
```
Comment (lieut-data) [inline app/plugin_install.go]: Is this comment adding value?

File: app/plugin_install.go [hunk 8]
```diff
@@ -103,53 +118,52 @@ func (a *App) installPluginLocally(pluginFile io.ReadSeeker, replace bool) (*mod
...
-	if err := a.notifyPluginStatusesChanged(); err != nil {
-		mlog.Error("failed to notify plugin status changed", mlog.Err(err))
+	// Activate plugin if it was previously activated.
+	pluginState := a.Config().PluginSettings.PluginStates[manifest.Id]
+	if pluginState != nil && pluginState.Enable {
+		updatedManifest, _, err := pluginsEnvironment.Activate(manifest.Id)
+		if err != nil {
+			return nil, model.NewAppError("installPluginLocally", "app.plugin.restart.app_error", nil, err.Error(), http.StatusInternalServerError)
+		}
+		manifest = updatedManifest
```
Comment (lieut-data) [inline app/plugin_install.go]: To confirm, this is new in the diff -- returning the updated manifest -- what do we expect to have changed to return, exactly?

File: app/plugin_install.go [hunk 9]
```diff
@@ -160,6 +174,11 @@ func (a *App) RemovePlugin(id string) *model.AppError {
 }
 
 func (a *App) removePlugin(id string) *model.AppError {
+	// Disable plugin before removal and notify cluster peers sync.
```
Comment (lieut-data) [inline app/plugin_install.go]: Could we elaborate on the `disable` part here? It's obvious that we're disabling, but it's not obvious why.

Also, do we need the comment re: notifying clusters sync? Seems like this comment could be part of the `DisablePlugin` godoc if needed.

File: app/plugin_install.go [hunk 10]
```diff
@@ -177,6 +196,7 @@ func (a *App) removePlugin(id string) *model.AppError {
 		return model.NewAppError("removePlugin", "app.plugin.remove_bundle.app_error", nil, err.Error(), http.StatusInternalServerError)
 	}
 
+	// Notify cluster peers async.
```
Comment (lieut-data) [inline app/plugin_install.go]: Is this comment adding value?

File: app/plugin_install.go [hunk 11]
```diff
@@ -212,24 +232,37 @@ func (a *App) removePluginLocally(id string) *model.AppError {
...
-	err = os.RemoveAll(pluginPath)
-	if err != nil {
+	if err := os.RemoveAll(pluginPath); err != nil {
 		return model.NewAppError("removePlugin", "app.plugin.remove.app_error", nil, err.Error(), http.StatusInternalServerError)
 	}
 
 	return nil
 }
 
+func (a *App) notifyPluginEvents(manifest *model.Manifest) error {
```
Comment (lieut-data) [inline app/plugin_install.go]: It's not entirely clear what `notifyPluginEvents` means, as a name -- there are so many "notify*plugin" invocations now.

**Comment by lieut-data** [comment]
Adding @crspeller for an extra set of eyes.

File: app/plugin_install.go (hunk 10 above)
Comment (ali-farooq0) [reply app/plugin_install.go]: just so the reader knows that this is an async call vs sync as we have both cases.

File: app/plugin.go (hunk 4 above)
Comment (crspeller) [reply app/plugin.go]: +1

File: app/plugin.go [hunk 12]
```diff
@@ -394,3 +398,48 @@ func (a *App) GetPlugins() (*model.PluginsResponse, *model.AppError) {
...
+
+		localStatus, err := a.GetPluginStatus(manifest.Id)
+		if err != nil {
+			return err
+		}
+		statuses = append(statuses, localStatus)
+
+		for _, status := range statuses {
+			// Count consistent peers
+			isPeerConsistent := status.Version == manifest.Version &&
```
Comment (crspeller) [inline app/plugin.go]: It seems like this might be unstable. We are checking parameters that are not directly relevant to our goal. All we really want to check is that all active nodes have copied over the file. Maybe we can create a flag just for that? Otherwise if the plugin happens to crash or the versions are incompatible this will break.

File: app/plugin.go [hunk 13]
```diff
@@ -394,3 +398,48 @@ func (a *App) GetPlugins() (*model.PluginsResponse, *model.AppError) {
...
+			if status.PluginId == manifest.Id && isPeerConsistent {
+				actualPeers++
+			}
+		}
+
+		// Notify all cluster peer clients.
+		if actualPeers == expectedPeers {
+			message := model.NewWebSocketEvent(model.WEBSOCKET_EVENT_PLUGIN_ENABLED, "", "", "", nil)
+			message.Add("manifest", manifest.ClientManifest())
+			a.Publish(message)
```
Comment (crspeller) [inline app/plugin.go]: This can happen multiple times because we are not making sure we are the last one to be in the correct state. We just make sure that everyone else is.  I don't think this is a problem though.

File: app/plugin.go (hunk 4 above)
Comment (ali-farooq0) [reply app/plugin.go]: funny enough, i had this flipped initially, but switched to this to guard against installing a plugin and then quickly enabling it, before the other peer has a chance to install it. In which case, it will only find the plugin for this peer and end up broadcasting to all clients, which may or may not result in a bundle 404.

However, even with this guard in place, I found hitting the `enable` endpoint returning a 500 because the peer i ended up calling `enable` on didn't have the plugin installed (yet), so don't think this is a full working solution. 

Since this race condition will most likely occur in a case where one is programmatically installing a plugin and then enabling it, i'm inclined to revert back to the original solution i had and document the case it's still failing to guard against.

File: app/plugin.go (hunk 12 above)
Comment (ali-farooq0) [reply app/plugin.go]: > All we really want to check is that all active nodes have copied over the file.

copied over the file on to the local FS?

what we really want to check is whether the plugin is enabled or not, as that's when we generate the webapp bundle. It is possible that the plugin is installed (files are copied over on to the local FS) but the webapp bundle isn't extracted, which will result in a 404 if clients reach out to the server asking for the plugin webapp bundle.

File: app/plugin.go (hunk 13 above)
Comment (ali-farooq0) [reply app/plugin.go]: True. I've updated webapp to make `plugin_enabled` and `plugin_disabled` events idempotent, to guard against multiple events going out.
https://github.com/mattermost/mattermost-webapp/pull/3425

File: app/plugin.go (hunk 12 above)
Comment (ali-farooq0) [reply app/plugin.go]: @crspeller 
but i do agree that this will be flaky, how about we check the webapp bundle hash?

File: app/plugin.go (hunk 3 above)
Comment (ali-farooq0) [reply app/plugin.go]: n/a. since flipping the logic to not notify when peers are not consistent.

File: app/plugin_install.go (hunk 6 above)
Comment (ali-farooq0) [reply app/plugin_install.go]: True, it's just a helper to combine `notifyPluginStatusesChanged` and `notifyPluginEnabled`. Got rid of it, moved the `IsActive` inside `notifyPluginEnabled`.

File: app/plugin_install.go (hunk 7 above)
Comment (ali-farooq0) [reply app/plugin_install.go]: since we have both async (`SendClusterMessage`) and sync (`ConfigChanged`), versions, thought it'd would help to clarify when writing code not to depend on the operation being sync.

File: app/plugin_install.go (hunk 8 above)
Comment (ali-farooq0) [reply app/plugin_install.go]: Ahh this was another gotcha i ran into. When activating a plugin, there is code that updates the manifest in memory that has the updated webapp bundle hash. Done here: https://github.com/mattermost/mattermost-server/blob/c1c2da804f088ba754b3fdce24caca4a697d4a18/plugin/environment.go#L246

Whereas fetching manifest through `pluginsEnvironment.Available()` even after a plugin is enabled, does not have the up to date webapp bundle hash as that is only in memory and not persisted to disk, which is why i have to do this, rather than just getting the upto date manifest

File: app/plugin_install.go (hunk 11 above)
Comment (ali-farooq0) [reply app/plugin_install.go]: removed

**Comment by ali-farooq0** [comment]
> Two other higher-level comments. First, I'd love to see a high-level description of the plugin activation/notification flow as you understand it put somewhere. Maybe as a top-level godoc comment in a `app/plugin_install.go`?

👍 

> Secondly, while I think I can convince myself that the convergence model on plugin statuses should work in most cases, if a node goes offline at the right time, I could see a scenario where no one is left to "finish". This might be a separate improvement ticket, but I wonder if we could eliminate the race condition serving up the bundle from server that hasn't yet unpacked it by routing that particular request all the way back to the file store? There's the typical thundering herd problem, plus the likely need to validate the hash we recover from the requested path, but throwing it out there for more discussion.

I do like the idea of unpacking the bundle on cache miss, however there are two race conditions that we're dealing with here.

1. Install plugin and immediately enable it
 - This won't be fixed by the suggestion above, as the peer won't even know if that plugin is enabled.

2. Install plugin. Wait. Enable plugin.
 - Since webapp bundle is unpacked on activation, the suggestion will mostly address this race condition, if we go with a variation of unpacking on installation rather than activation.

I did implement a version where I was generating webapp bundle on installation, it turned out a bit a bit messy, as `Environment` needed to expose an api to generate and cleanup webapp bundles as only it has access to `env.webappPluginDir` and `env.pluginDir`. Maybe we go with that approach for now to eliminate race on unpacking on activation?

Thoughts @lieut-data?

File: app/plugin_install.go [hunk 14]
```diff
@@ -1,6 +1,22 @@
 // Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
 // See LICENSE.txt for license information.
 
+// Following outlines the plugin installation/activation flow.
```
Comment (lieut-data) [inline app/plugin_install.go]: I think this captures the technical details, but it might be a bit hard to follow without a lot of domain context. Taking a different stab at this:

> Installing a managed plugin consists of copying the uploaded plugin (*.tar.gz) to the filestore, unpacking to the configured local directory (PluginSettings.Directory), and copying any webapp bundle therein to the configured local client directory (PluginSettings.ClientDirectory). The unpacking and copy occurs each time the server starts, ensuring it remains synchronized with the set of installed plugins. 

> When a plugin is enabled, all connected websocket clients are notified so as to fetch any webapp bundle and load the client-side portion of the plugin. This works well in a single-server system, but requires careful coordination in a high-availability cluster with multiple servers. In particular, websocket clients must not be notified of the newly enabled plugin until all servers in the cluster have finished unpacking the plugin, otherwise the webapp bundle might not yet be available. Ideally, each server would just notify its own set of connected peers after it finishes this process, but nothing prevents those clients from re-connecting to a different server behind the load balancer that hasn't finished unpacking.

> To achieve this coordination, each server instead checks the status of its peers after unpacking. If it finds peers with differing versions of the plugin, it skips the notification. If it finds all peers with the same version of the plugin, it notifies all websocket clients connected to all peers. There's a small chance that this never occurs if the the last server to finish unpacking dies before it can announce. There is also a chance that multiple servers decide to notify, but the webapp handles this idempotently.

> Complicating this flow further are the various means of notifying. In addition to websocket events, there are cluster messages between peers. There is a cluster message when the config changes and a plugin is enabled or disabled. There is a cluster message when installing or uninstalling a plugin. There is a cluster message when peer's plugin change its status. And finally the act of notifying websocket clients is propagated itself via a cluster message.

> The key methods involved in handling these notifications are notifyPluginEnabled and notifyPluginStatusesChanged. Note that none of this complexity applies to single-server systems or to plugins without a webapp bundle.

> Finally, in addition to managed plugins, note that there are unmanaged and prepackaged plugins. Unmanaged plugins are plugins installed manually to the configured local directory (PluginSettings.Directory). Prepackaged plugins are included with the server. They otherwise follow the above flow, except do not get uploaded to the filestore. Prepackaged plugins override all other plugins with the same plugin id. Managed plugins override unmanaged plugins with the same plugin id.

File: app/plugin_install.go (hunk 7 above)
Comment (lieut-data) [reply app/plugin_install.go]: I admit the comment doesn't really help capture the nuance of those two methods -- that is, nothing here tells me about `ConfigChanged` and why I care that this is async vs. that being sync.

**Comment by ali-farooq0** [comment]
> Thanks, @ali-farooq0! To confirm, are we still unpacking the webapp bundles for unmanaged/prepackged plugins? (I think that follows from the local install method, but just wanted to clarify.)
> 
> A few comments below.

Yes, it will continue to unpack the webapp bundle on activation for unmanaged/prepackaged plugins, same as before. The new addition is to unpack plugins on installation.

File: app/plugin_install.go (hunk 14 above)
Comment (ali-farooq0) [reply app/plugin_install.go]: ❤️

File: app/plugin.go [hunk 15]
```diff
@@ -394,3 +399,55 @@ func (a *App) GetPlugins() (*model.PluginsResponse, *model.AppError) {
...
+	if pluginsEnvironment == nil {
+		return errors.New("pluginsEnvironment is nil")
+	}
+	if !manifest.HasClient() || !pluginsEnvironment.IsActive(manifest.Id) {
+		return nil
+	}
+
+	var statuses model.PluginStatuses
+
+	if a.Cluster != nil {
```
[the remark this answers is not in the document]
Comment (ali-farooq0) [reply app/plugin.go]: Can't return early when `a.Cluster == nil` because this needs to notify websocke clients in non-HA mode.

File: app/plugin.go [hunk 16]
```diff
@@ -394,3 +399,55 @@ func (a *App) GetPlugins() (*model.PluginsResponse, *model.AppError) {
...
+		return err
+	}
+	statuses = append(statuses, localStatus)
+
+	// This will not guard against the race condition of enabling a plugin immediately after installation.
+	// As GetPluginStatuses() will not return the new plugin (since other peers are racing to install),
+	// this peer will end up checking status against itself and will notify all webclients (including peer webclients),
+	// which may result in a 404.
+	for _, status := range statuses {
+		if status.PluginId == manifest.Id && status.Version != manifest.Version {
```
[the remark this answers is not in the document]
Comment (ali-farooq0) [reply app/plugin.go]: unfortunately in that case it'll notify incorrectly. Ideally I'd want to compare the webapp bundle hash as that's what accurately determines if anything has changed for webapp. But i only have access to my bundle hash and not of other peers :(

File: app/plugin_install.go [hunk 17]
```diff
@@ -1,6 +1,39 @@
...
+//
+// The key methods involved in handling these notifications are notifyPluginEnabled and notifyPluginStatusesChanged.
+// Note that none of this complexity applies to single-server systems or to plugins without a webapp bundle.
+//
+// Finally, in addition to managed plugins, note that there are unmanaged and prepackaged plugins.
+// Unmanaged plugins are plugins installed manually to the configured local directory (PluginSettings.Directory).
+// Prepackaged plugins are included with the server. They otherwise follow the above flow, except do not get uploaded
+// to the filestore. Prepackaged plugins override all other plugins with the same plugin id. Managed plugins
+// override unmanaged plugins with the same plugin id.
+//
```
[the remark this answers is not in the document]
Comment (ali-farooq0) [reply app/plugin_install.go]: That was all @lieut-data. Huge shoutout to @lieut-data for fixing up the docs  🎉

File: app/plugin.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (marianunez) [reply app/plugin.go]: Gotcha!

File: app/plugin.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (marianunez) [reply app/plugin.go]: Reinstalling with the same version number is a rare case considering that plugins release bumping their versions accordingly. I would only see it being more common in a development environment where we have to be mindful on realoading the webapp and should not be an issue.

[the remark this answers is not in the document]
Comment (ali-farooq0) [reply app/plugin.go]: true, however i know someone mentioned that community folks sometimes release with the same version :( though they should be bumping the version . Fixing the root cause would be the best way to completely addressing the issue, which would be to expose the bundle hash through status somehow.

[the remark this answers is not in the document]
Comment (marianunez) [reply app/plugin.go]: Maybe add that to the spike where we are collecting the future improvements ?

[the remark this answers is not in the document]
Comment (ali-farooq0) [reply app/plugin.go]: Done. https://mattermost.atlassian.net/browse/MM-17871

**Comment by ali-farooq0** [comment]
@crspeller **poke** for final review :)

**Comment by ali-farooq0** [comment]
Removing QA label as @prapti as already approved https://github.com/mattermost/mattermost-webapp/pull/3425#pullrequestreview-278562185

**Comment by amyblais** [comment]
@aaronrothschild Please help check if this needs documentation for v5.15.

**Comment by aaronrothschild** [comment]
Docs submitted for PR here: https://github.com/mattermost/docs/pull/2989
